### ISSUE TITLE
Memory interfaces

### ISSUE THREAD

**Issue by aehart** [cms-tracklet/firmware-hls#58]
I sort of mentioned this in passing during the meeting today, but the changes that @djcranshaw and @bryates made to the memory interfaces got me thinking about other ports that are actually redundant now that we are plugging everything into a Verilog top module:

- `nentries` ports for output memories: We don't really need these because we can just count the number of write enables in the top module. These can be removed by removing the `nentries_` data member from the memory classes for output memories. It was easy to make this change locally, and it has the desired effect. I can submit a pull request soon if we decide this is OK.
- `nentries` ports for each page of the input memories: We can make the processing modules "BX agnostic" such that they have no BX input port and have a single `nentries` port for each input memory, instead of one for each memory page. Then it will be up to the top module to read from and write to the correct page without unnecessarily routing that information through all the processing modules. I haven't attempted this change yet.

Any thoughts?

**Comment by djcranshaw** [comment]
- On the first point. Would this require a change to Memory.v? I think it has to because the write-enables and data are wired directly into the memory module. This is still probably doable, just means logic internal to Memory.v Another thing is that we need the nentries_ array for input memories, so we can't just get rid of the data member without separating input memories from output memories in the HLS. We would probably need something like MemoryTemplateInputs.hh and MemoryTemplateOutputs.hh, where the former has nentries_ and the latter doesn't. That strikes me as a bit of a nuisance but it still would work. Another possibility is that if we don't ever use the nentries_ array in output memories, HLS will just optimize those ports away. Then we could still just have the one MemoryTemplate.hh file. I haven't tried this but it this should work, since HLS successfully optimized away the analogous input ports when we stopped using them.

- On the second point. So the way it is set up right now is that the processing blocks handle the bx signals, and the memories are bx-agnostic. I like this setup because after a proc block is finished with a processing step, the last thing it does is send out the bx_signal it was working on, coincident with its done signal. Then the next block knows to start at the same moment its getting the information about which bx its working on. If we make the proc blocks bx-agnostic, we will need some verilog logic in the top module to keep track of when each module is finished, so it can switch the bx signal (which would probably have to be a register array in the top-level or something) that is feeding the output memory module so that it knows which page to write to. Also, if we only have one nentries input per proc block, we will have to dynamically switch which page in the memory the proc block is reading from, which will also require some verilog logic in the top-level. My preference would be keeping as much logic in the HLS as possible.

**Comment by gershtein** [comment]
Hi Derek,

On Nov 20, 2019, at 10:51 AM, Derek Cranshaw <notifications@github.com<mailto:notifications@github.com>> wrote:


  *   On the second point. So the way it is set up right now is that the processing blocks handle the bx signals, and the memories are bx-agnostic. I like this setup because after a proc block is finished with a processing step, the last thing it does is send out the bx_signal it was working on, coincident with its done signal. Then the next block knows to start at the same moment its getting the information about which bx its working on. If we make the proc blocks bx-agnostic, we will need some verilog logic in the top module to keep track of when each module is finished, so it can switch the bx signal (which would probably have to be a register array in the top-level or something) that is feeding the output memory module so that it knows which page to write to. Also, if we only have one nentries input per proc block, we will have to dynamically switch which page in the memory the proc block is reading from, which will also require some verilog logic in the top-level. My preference would be keeping as much logic in the HLS as possible.


But the module start / stop for events are not dynamic, right? All modules work for 450 ns and then switch events.
So I’m not sure why you say we should track when the module is finished?

-y

**Comment by aehart** [comment]
@djcranshaw 

- Yes, this would require some additional logic in Memory.v to increment the number of entries whenever something is written. On the C++ side, the way I implemented it is by removing `nentries_` from the current `MemoryTemplate` and `MemoryTemplateBinned`, and putting it in derived classes that inherit from the original classes, which I called `InputMemoryTemplate` and `InputMemoryTemplateBinned`. Likewise, I moved the `getEntries` and `clear` methods to the derived classes. Then in each of the specific memory classes, I just have to define an "input" version; e.g., in VMStubTEInnerMemory:
```c++
// Memory definition
template<int VMSTEIType> using VMStubTEInnerMemory =  MemoryTemplate<VMStubTEInner<VMSTEIType>, 1, kNBits_MemAddr>;
template<int VMSTEIType> using VMStubTEInnerInputMemory =  InputMemoryTemplate<VMStubTEInner<VMSTEIType>, 1, kNBits_MemAddr>;
```
- There seems to be some inconsistency regarding the BX ports on the processing modules. The TrackletEngine and TrackletCalculator both have just a single BX input, while I see now that the ProjectionRouter has an output port as well. So maybe we should discuss what these ports should look like, but I think I might see your point.</br>
And yes, with a single `nentries` port, we'd have to dynamically switch which `nentries` is being fed into the module. My inclination would be to make the memory interfaces as simple as possible and move that logic to the Verilog top module, but maybe at this point, we should just keep the BX port(s) and input memory interfaces as they are, and keep this possibility in mind if we really _need_ to optimize the resource utilization or timing in the future.

**Comment by djcranshaw** [comment]
@aehart
The InputMemoryTemplate.hh inheriting from MemoryTemplate.hh solution looks nice to me, if we decide to make this change this looks like the natural way to do it. Incidentally, I tried just removing the line which updates the nentries_ array in write_mem in the hopes that HLS would realize that the member isn't being used anywhere and just optimize the port away. Instead, it threw some warnings at me and auto-generated an input port for it. Oh well.

@gershtein
It's true that the start/stop aren't dynamic, but because of the FIFO latency of the individual modules, they need not start all at the same time. For example, in a PR-ME-MC chain, if PR starts working on BX=0 at t=0, and it has a FIFO latency of, say, 6 clocks, ME shouldn't start working on BX=0 until clock 114 = 475 ns (@240 MHz). Then PR starts working on BX=1 at 450 ns, and ME starts working on BX=1 at 475+450 = 925 ns. If the processing blocks don't send out a BX signal, we either have to manually adjust the top-level verilog / project generation scripts whenever a change in a processing block causes a change in its FIFO latency, or have the top-level module listen for a done signal from the proc blocks, and adjust a register array in the top-level, which would then control which nentries port in the mem module that the single remaining nentries port in the proc module is reading from or writing to.

**Comment by gershtein** [comment]
Hi Derek,

right, so not everything starts at the n*450ns. But everything still starts at pre-determined times, correct?
I.e. you do not try to gain some processing time by borrowing from the next event?
In that case, you can still have a “conductor” module at the top that sends out signals when it’s time to switch events.
I think it’s preferable to having them in the modules themselves, but sure, it can work either way.

cheers,
-y

On Dec 3, 2019, at 10:24 PM, Derek Cranshaw <notifications@github.com<mailto:notifications@github.com>> wrote:


@aehart<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Faehart&data=02%7C01%7Cygershte%40physics.rutgers.edu%7C6768d0013d8f439b2f0608d778372f6b%7Cb92d2b234d35447093ff69aca6632ffe%7C1%7C0%7C637110050731249085&sdata=DCJGPZk6Kcfdv2glgWqHqMZX0SC5N1hXcHcDJhuTtRM%3D&reserved=0>
The InputMemoryTemplate.hh inheriting from MemoryTemplate.hh solution looks nice to me, if we decide to make this change this looks like the natural way to do it. Incidentally, I tried just removing the line which updates the nentries_ array in write_mem in the hopes that HLS would realize that the member isn't being used anywhere and just optimize the port away. Instead, it threw some warnings at me and auto-generated an input port for it. Oh well.

@gershtein<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgershtein&data=02%7C01%7Cygershte%40physics.rutgers.edu%7C6768d0013d8f439b2f0608d778372f6b%7Cb92d2b234d35447093ff69aca6632ffe%7C1%7C0%7C637110050731249085&sdata=p7kc%2BbyczBbN6E%2FoJ2RoK5y%2BqGUyp1eCHObP2fde8vE%3D&reserved=0>
It's true that the start/stop aren't dynamic, but because of the FIFO latency of the individual modules, they need not start all at the same time. For example, in a PR-ME-MC chain, if PR starts working on BX=0 at t=0, and it has a FIFO latency of, say, 6 clocks, ME shouldn't start working on BX=0 until clock 114 = 475 ns (@240<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2F240&data=02%7C01%7Cygershte%40physics.rutgers.edu%7C6768d0013d8f439b2f0608d778372f6b%7Cb92d2b234d35447093ff69aca6632ffe%7C1%7C0%7C637110050731259043&sdata=mGycF3pgaoaPNxdn4NW7zwPGSVV8aJuitl4m8hyMWbk%3D&reserved=0> MHz). Then PR starts working on BX=1 at 450 ns, and ME starts working on BX=1 at 475+450 = 925 ns. If the processing blocks don't send out a BX signal, we either have to manually adjust the top-level verilog / project generation scripts whenever a change in a processing block causes a change in its FIFO latency, or have the top-level module listen for a done signal from the proc blocks, and adjust a register array in the top-level, which would then control which nentries port in the mem module that the single remaining nentries port in the proc module is reading from or writing to.

**Comment by pwittich** [comment]
where are we with this issue? Is it still live?

**Comment by aehart** [comment]
I was going to submit a PR with the `nentries` ports removed for output memories. I got this to work in terms of what HLS synthesizes, but I spent ~way too long~ quite some time trying to get the C test benches to work with this change in the memory interfaces, and ultimately, I don't think this is possible.

The problem is that if you remove the `nentries_` data member from the class for output memories, then you have no reliable way to count the number of entries in the test bench. In the top level HDL, you can just count the number of write-enables for that memory, but you can't really do that in the C test bench. I think we want to retain the ability in the test benches to check the number of entries that were written to each output memory in addition to checking the contents of the memories. And I don't think there's a simple way to do this without keeping the `nentries_` data member, even though it leads to redundant ports in the exported IP cores.

So I think that concludes this issue, unless someone else has any comments.

**Comment by djcranshaw** [comment]
Here is something that might work. If we leave the nentries_ member in the output memories, and then modify our write_mem function along the following lines:

bool write_mem(BunchXingT ibx, DataType data, int addr_index)
  {
#pragma HLS ARRAY_PARTITION variable=nentries_ complete dim=0
#pragma HLS inline
    if (addr_index <= (1<<NBIT_ADDR)) {
      dataarray_[ibx][addr_index] = data;
#ifndef __SYNTHESIS__
      nentries_[ibx] = addr_index + 1;
#endif
      return true;
    } else {
      return false;
    }
  }

This makes for terribly unreadable code, and might also be starting us down a slippery slope to which we don't want to venture too close. But it should work, in that the C-test bench has an easy way to check the number of entries in the output, and the HLS won't synthesize the nentries ports in output memories. By the way, earlier in this thread I said that HLS synthesizes input ports for members that aren't used, and throws warnings at you. Indeed it does, unless you give it the tcl command that trims dangling ports, in which case it doesn't synthesize those ports (and doesn't throw warnings).

So what do people think? Is it worth putting ifdefs in the middle of our functions to save having to specify the number of entries at the start of our firmware chains, as well has having needless nentries input ports to our memories in firmware?

**Comment by aehart** [comment]
I tried something similar with the `SYNTHESIS` pragma (though not exactly this), and it worked fine with C-simulation and C-synthesis, but the C/RTL cosimulation didn't like it _at all_. Without actually trying out your solution, I would guess that you'd run into similar problems.

My opinion is that trying to get rid of these redundant `nentries` ports is just not worth the trouble at this point.

**Comment by djcranshaw** [comment]
Hi all. So I spent a bit of time these last two weeks looking into this and I figured I'd resurrect this thread to get some discussion going on the things I've found out. I'll use ProjectionRouter as the concrete example because I've been working with it a lot recently and I understand it fairly well.

The Model
This particular Projection Router (PR) has 8 Tracklet Projections (tproj) inputs and an All Projection (aproj) and 8 VMProjections (vmproj) as outputs on the interface of the top-level HLS. Each one of these memories is a class with several constructors, getters/setters etc. and two members:
An array called dataarray, which stores the actual data in the intermediate BRAM and which HLS synthesizes as a BRAM interface
An array called nentries, which stores the number of entries on each page of the BRAM, and which HLS synthesizes as separate data ports (input ports for the tproj, output for the aproj and vmproj) because we tell it to using the partition pragma.
In the top level HDL, the ports that the HLS synthesizes for dataarray and nentries are wired directly into memory.v, which stores both the data and the number of entries for each page.

The Problem
It is straightforward (and I think Andrew has already done this) to modify our memory.v so that, instead of relying on the PR to tell the memory.v instance how many entries are on each page, the memory.v instance counts the number of writes to itself and stores the number of entries internally that way. Doing this would save us the output nentries ports on the interface of the PR, and any associated logic inside the PR needed to write to those ports. The problem is that the HLS test bench (at least the way it is currently set up) needs to know how many entries a page has in order to compare it to the reference file to make sure the behavior of the PR is correct. If we don't have an nentries array, or something equivalent, coming out of the top-level HLS, there is no way to get that information. So it seems that, if we want to know how many writes to memory were performed in C-Sim, we need to pass that information out of the top-level to the test bench, which means a port will be synthesized.

There are several options for how to deal with this.

Option 1, don't check nentries for the top-level HLS
One possibility would be to simply not check how many entries the HLS is writing, and just have the HLS test bench check the output memory against the contents of the memory in the reference file, up to the number of entries in the reference file. This allows us to keep the simulation and synthesis identical and the c/rtl cosimuation would continue to work. This has two unappealing properties though
1. We would fail to catch errors where the HLS firmware wrote additional memories than what was expected by the reference file.
2. We would need something to tell the memory.v to reset its memory every bunch crossing, either a controlling signal in the top-level or an internal counter in memory.v itself. We can't rely on the HLS blocks to just tell memory.v where to write and have nentries updated that way because there is no guarantee a particular memory will be written to on a particular bx
Another inconvenience of this method is that certain kinds of errors might look like others, making debugging difficult. For example, if on bx0 the FW worked correctly and there were 4 entries and the 4th entry was 0xABCD, then on BX2 there was a bug and the HLS only wrote three entries but the ref file expects 4, the error reported would be an inconsistency in slot 4 (0xABCD instead of whatever it was supposed to be), rather than a missing entry. And we could get very lost trying to figure out how on Earth the HLS managed to get 0xABCD for those particular inputs.

Option 2, keep nentries for C-SIM, but get rid of them during synthesis using the __SYNTHESIS__ pragma
Another possibility is that we do exactly what we're doing now, but we guard the nentries writes with an #ifndef __SYNTHESIS__. The advantage here is that we can do a fully correct C-Simulation checking for all the errors that we currently check for, and our synthesized FW won't have the nentries ports, and as long as we can count on our memory.v to keep track of the nentries correctly by counting the FW should behave as we expect. The problem here is that the C/RTL cosimulation would definitely fail, because the synthesized test bench would try to read the nentries ports that it thinks should be there but won't be because of the pragma. This option would be more appealing if we hadn't seen so many cases in the past where csim passes but c/rtl cosim fails. I'm reluctant to give up on our insistence that the HLS passes cosim, but it is a possibility.

Option 3, keep nentries, but redesign the HLS so that they don't use any extra FW.
This possibility to me is the most appealing if it works, because we don't have to restructure our top-level or HLS interfaces, just what's inside them. First, a reminder that several months ago, the HLS modules made the switch to keeping track of the number of entries internally for a particular bx, to avoid having to read in the number of memories as an input port from the output memory. So the reasoning is that since we keep track of this number internally anyways, it should cost no (or very-nearly no) extra logic to just wire that number to an interface port. I managed to do this for the PR, the main difference was to notice that it is easier to write the number of entries for every memory on every loop iteration unconditionally, rather than clearing the memories at the start and only updating a particular memory nentries when it is written to. This seems a bit strange from a SW perspective but it kinda makes sense in FW; you save all the logic you would otherwise need to check whether you should write to that port or not, and you save all the logic you would need to ensure that you zero all the memories before the processing loop starts. After implementing this in PR the HLS resource usage was exactly 5 FF more (i.e. pretty much the same) than for the PR with all references to nentries removed and no nentries ports synthesizes. I haven't tried this for other modules but I see no reason this same strategy couldn't be implemented for all the other modules. This has all the advantages in that cosim will continue to work, we catch all the same errors we're catching now, and the FW footprint is the same as if there were no nentries ports at all. One possible downside is that if we fail to remove the nentries ports, we may run into routing issues as the project gets large. I don't really understand the details of that but it's perhaps something to keep in mind.

There may be other options too (I briefly considered reserving one slot in the BRAM after the 108 that the algo uses to write the nentries of the page, and have our memory.v listen for writes to that address and instead route them to an internal nentries register, but I think that way madness lay for both HDL and HLS), but these are what I have for now. I could try to put together a talk to better explain these options with diagrams and code snips if its useful, but I won't be able to do that in time for today's talk and I didn't want to wait 2-weeks before talking about this (also today's agenda is a bit full anyways). Tagging @aehart  because I know he worked on this in the past but any comments and suggestions from anyone are welcome.

**Comment by pwittich** [comment]
What is the status of this issue?

**Comment by djcranshaw** [comment]
This is the issue opened to get rid of the nentries ports for the output memories attached to each HLS block. To do that we've decided to adopt possibility 2 from slide 6 in this talk https://indico.cern.ch/event/940922/contributions/3955291/attachments/2079826/3493207/HLSChat_nentries_2020-07-24.pdf
The work for that is already done on branch RemoveNEntries, but since that branch has touched several files and the several PRs since then have done so as well, I'm anticipate a few merge conflicts that will need to be resolved. Once fw_synch is merged into master, I will perform the rebase and open the PR from RemoveNEntries. Once that gets merged, this issue can be closed.

**Comment by pwittich** [comment]
thanks @djcranshaw

**Comment by pwittich** [comment]
@djcranshaw The PR #114 has been merged so please make your new PR for this issue ASAP
